test-runner-poc/lib/runner.js

```diff
@@ -1,73 +1,143 @@
+      }
+      environmentCallback(environment, client);
+    }
+  });
+}
+
+function createFilters(callback) {
+  let filtersInitialized = 0;
+  const filterFiles = fs.readdirSync(path.join(__dirname, 'filters'));
+
+  let topology, version;
 
-	const environmentName = process.env['MONGODB_ENVIRONMENT'];
+  filterFiles.filter(x => x.indexOf('js') !== -1).forEach(x => {
+    const FilterModule = require(path.join(__dirname, 'filters', x));
+    const filter = new FilterModule();
 
-	console.log(`[environment: ${environmentName}]`);
+    if (typeof filter.initializeFilter === 'function') {
+      filter.initializeFilter(_increment);
+    } else {
+      _increment();
+    }
 
-	//apply filters
-	fs.readdirSync(path.join(__dirname, "filters"))
-		.filter(x => x.indexOf("js") !== -1)
-		.forEach(x => {
-			const FilterModule = require(path.join(__dirname, "filters", x));
-			addFilter(new FilterModule({ runtimeTopology: 'sharded' }));
-		});
+    topology = topology || filter.runtimeTopology;
+    version = version || filter.mongoVersion;
 
+    //Makes sure to wait for all the filters to be initialized and added before calling the callback
+    function _increment() {
+      //as filters finish executing initializeFilter, this will increase
+      filtersInitialized += 1;
+      addFilter(filter);
+
+      //after all filters have been initialized, further environment handling is able to begin
+      if (filtersInitialized === filterFiles.length) {
+        callback(topology, version);
+      }
+    }
+  });
+}
+
+before(function(done) {
+  environmentSetup((environment, client) => {
+    this.configuration = new TestConfiguration(environment);
+    client.close(done);
+  });
 });
 
-beforeEach(function() {
-	var self = this;
-
-	var called = 0;
-	function callback() {
-		called += 1;
-		if (called === filters.length) _run();
-	}
-
-	if (filters.length) {
-		filters.forEach(function(filter) {
-			callback();
-		});
-	}
-
-	function _run() {
-		if (!applyFilters(self.currentTest)) {
-			self.skip();
-		}
-	}
+beforeEach(function(done) {
+  // Assigned this to a new variable called self in order to preserve context and access tests within the _run function.
+  const self = this;
+  let filtersExecuted = 0;
+  if (filters.length) {
+    filters.forEach(function(filter) {
+      _run(filter);
+    });
+  }
+
+  function _run(filter) {
+    filtersExecuted += 1;
+
+    if (!filter.filter(self.currentTest)) {
+      if (!self.currentTest.parent.parent.root) {
+        // self.currentTest.parent.pending = true; <-- this makes apm_tests skip when they should not
+        self.currentTest.parent._beforeEach = [];
+      }
+      self.skip();
+    }
+    if (filtersExecuted === filters.length) {
+      done();
+    }
+  }
 });
 
-function applyFilters(test) {
-	return filters.every(function(filterFunc) {
-		var res = filterFunc.filter(test);
-		return res;
-	});
-}
+afterEach(() => mock.cleanup());
+
+after(function(done) {
+  mongoClient.close(() => {
```

Comment:
Why is this `mongoClient.close(() => { done(); });` instead of `mongoClient.close(done);`?